acme/dns_challenge.go

```diff
@@ -60,50 +61,129 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 		}
 	}()
 
-	fqdn, _, _ := DNS01Record(domain, keyAuth)
+	fqdn, value, _ := DNS01Record(domain, keyAuth)
 
-	preCheckDNS(domain, fqdn)
+	err = waitFor(60, 2, func() (bool, error) {
+		return preCheckDNS(domain, fqdn, value)
+	})
 
 	return s.validate(s.jws, domain, chlng.URI, challenge{Resource: "challenge", Type: chlng.Type, Token: chlng.Token, KeyAuthorization: keyAuth})
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn, value string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
```

Comment:
I think this should be `lookupNameservers(toFqdn(fqdn))` instead.

Scenario: a request for `domain = "foo.bar.com"`, which would have `fqdn = "_acme-challenge.foo.bar.com"`.

Assuming `bar.com IN NS ns1.bar.com`, `lookupNameservers("foo.bar.com.")` would return `["ns1.bar.com"]`, and the function would query `ns1.bar.com` for `_acme-challenge.foo.bar.com`. This is the common case.

However, `ns1.bar.com` could have a record `_acme-challenge.foo.bar.com IN NS ns.confusion.party` – delegating *that specific name* to a different nameserver, in which case this function should query `ns.confusion.party` instead.

`lookupNameservers(toFqdn(fqdn))` would be able to find this `NS` record and follow the delegation, while `lookupNameservers(toFqdn(domain))` would not.